### ISSUE TITLE
"Uncaught Error: Attempted to remove a dependant that was no longer registered"

### ISSUE THREAD

**Issue by matthew-dean** [ractivejs/ractive#1995]
I see that there have been a couple issues opened and later closed for this error. This is something I'm seeing consistently for one of my templates. What are some things that might cause this?

**Comment by Madgvox** [comment]
I'm afraid we're going to need a bit more context. Can you reproduce it in jsFiddle, or paste the relevant template here?

**Comment by matthew-dean** [comment]
Arg, I tried to boil it down to a limited use case, but I can't reproduce the error. But I'm wondering if it's because the project list is being spliced and resorted elsewhere, and then it's triggering an update on the ractive. This is the non-error version: http://codepen.io/matthewdean/pen/QbMpKj

**Comment by matthew-dean** [comment]
Another Ractive error I'm seeing immediately after this one.

>  Error: Failed to execute 'insertBefore' on 'Node': The node before which the new node is to be inserted is not a child of this node.

Does it have anything to do with the project nodes removing and immediately re-inserting in a different spot, based on the open flag? Is there a way to determine what Ractive is trying to do here? Which node it's trying to insert / move?

**Comment by matthew-dean** [comment]
I just now noticed that the loading state sets incorrectly because the keypath is moved. Maybe another thing I could try: is there a way for Ractive to change the "display" order of the array without modifying the source array order? My suspicion is that the array splicing / moving is what's tripping up Ractive, as it may be losing a reference to the original object during an update. Do you agree?

**Comment by Madgvox** [comment]
> is there a way for Ractive to change the "display" order of the array without modifying the source array order? 

That is not something that is supported natively by ractive.

I don't know what to tell you, your ractive looks like it should be working perfectly, and it is from what I can tell. If you're only adding or removing nodes through ractive you shouldn't ever get the `insertBefore` error. And I don't see anything that would cause an unregistered dependant error. Although in your pen you're never adding or removing entries, which is what is the most likely cause of such errors.

The only change I made worth mentioning here is that I moved the `sort` onto the ractive method, just to make sure the change was registered properly:

```js
ractive.sort('projects', function(a, b) {
  if (a.open)
    return -1;
  if (b.open)
    return 1;

  return 0;
});
```

**Comment by Rich-Harris** [comment]
Looks like this has been languishing a while – @matthew-dean are you still experiencing this issue, or can it be closed?

**Comment by matthew-dean** [comment]
I ended up creating a workaround, which was to return a (lodash) filtered list in two separate each blocks. So, not the most efficient, but probably negligible since project lists aren't long. Therefore, I don't know if this was fixed. It seemed like the problem was when nodes were not updated, but effectively "moved" to another place in a node tree, which Ractive seemed to have a hard time reconciling. So, on an update, or subsequent updates, a node was moved from one place to another.
